File: 33_recursion/33_1_robot_instructions.cpp
#include <iostream>
#include <string>

std::string robot_instructions(std::string seq) {
  if (seq.find("2") == std::string::npos) {
    return seq;
  }

  if (seq.find("2") != std::string::npos && seq.length() == 2) {
    return seq.substr(1);
  }

  size_t index = seq.find("2");
  std::string first_half, second_half;

  if (index == 0) {
    first_half = seq.substr(1);
    second_half = seq.substr(2);
  } else {
    first_half = seq.substr(0, index);
    second_half = seq.substr(index);
  }

  return robot_instructions(first_half) + robot_instructions(second_half);
}

int main() {
  // Example1
  std::string seq = "LL";
  std::cout << robot_instructions(seq) << std::endl;

  // Example2
  seq = "2LR";
  std::cout << robot_instructions(seq) << std::endl;

  // Example3
  seq = "2L";
  std::cout << robot_instructions(seq) << std::endl;

  // Example4
  seq = "22LR";
  std::cout << robot_instructions(seq) << std::endl;

  // Example5
  seq = "LL2R2L";
  std::cout << robot_instructions(seq) << std::endl;

  return 0;
}

File: 33_recursion/33_1_robot_instructions.py
def robot_instructions(seq: str)->str:
  # Base Case1
  if "2" not in seq:
    return seq
  # Base Case2
  if "2" in seq and len(seq) == 2:
    return seq[1]

  index = seq.find("2")
  if index == 0:
    first_half = seq[1:]
    second_half = seq[2:]
  else:
    first_half = seq[:index]
    second_half = seq[index:]

  return robot_instructions(first_half) + robot_instructions(second_half)

if __name__ == "__main__":
  # Example1
  seq = "LL"
  print(robot_instructions(seq))
  # Example2
  seq = "2LR"
  print(robot_instructions(seq))
  # Example3
  seq = "2L"
  print(robot_instructions(seq))
  # Example4
  seq = "22LR"
  print(robot_instructions(seq))
  # Example5
  seq = "LL2R2L"
  print(robot_instructions(seq))